src/autora/state.py

```diff
@@ -1363,3 +1384,77 @@ def experiment_runner(conditions: pd.DataFrame, **kwargs):
         return Delta(experiment_data=experiment_data)
 
     return experiment_runner
+
+
+def combined_functions_on_state(
+    functions: List[Tuple[str, Callable]], output: Optional[Sequence[str]] = None
+):
+    """
+    Decorator (factory) to make target list of `functions` into a function on a `State`.
+    The resulting function uses a state field as input and combines the outputs of the
+    `functions`.
+
+    Args:
+        function: the list of functions to be wrapped
+        output: list specifying State field names for the return values of `function`
+
+    Examples:
+        >>> @dataclass(frozen=True)
+        ... class U(State):
+        ...     conditions: List[int] = field(metadata={"delta": "replace"})
+        >>> identity = lambda conditions : conditions
+        >>> double_conditions = combined_functions_on_state(
+        ...     [('id_1', identity), ('id_2', identity)], output=['conditions'])
+        >>> s = U([1, 2])
+        >>> s_double = double_conditions(s)
+        >>> s
+        U(conditions=[1, 2])
+        >>> s_double
+        U(conditions=[1, 2, 1, 2])
+
+        # We can also pass parameters to the functions:
+        >>> def multiply(conditions, multiplier):
+        ...     return [el * multiplier for el in conditions]
+        >>> double_and_triple = combined_functions_on_state(
+        ...     [('doubler', multiply), ('tripler', multiply)], output=['conditions']
+        ... )
+        >>> s = U([1, 2])
+        >>> s_double_triple = double_and_triple(
+        ...     s, params={'doubler': {'multiplier': 2}, 'tripler': {'multiplier': 3}}
+        ... )
+        >>> s_double_triple
+        U(conditions=[2, 4, 3, 6])
+
+        # If the functions return a Delta object, we don't need to provide an outpur argument
+        >>> def decrement(conditions, dec):
+        ...     return Delta(conditions=[el-dec for el in conditions])
+        >>> def increment(conditions, inc):
+        ...     return Delta(conditions=[el+inc for el in conditions])
+        >>> dec_and_ind = combined_functions_on_state(
+        ...     [('decrement', decrement), ('increment', increment)])
+        >>> s_dec_and_inc = dec_and_ind(
+        ...     s, params={'decrement': {'dec': 10}, 'increment': {'inc': 2}})
+        >>> s_dec_and_inc
+        U(conditions=[-9, -8, 3, 4])
```

Comment:
similar confusion here. doesn't `s = U([1, 2])` still? how/where does doubling the conditions come into play? also, why aren't both `dec` and `inc` applied to every element to obtain new values (given the specifications of `el-dec for el in conditions` and `el+inc for el in conditions` in the respective return values)?

the values `[-9, -8, 3, 4]` would result from 1-10=-9, 2-10=-8, 1+2=3, 2+2=4, correct? but why wouldn't it be 1-10+2=-7, 2-10+2=-6, ...?

sorry that i'm so confused!

Reply:
The idea is to have a "parallel" pipeline instead of a "serial". The functions in the list get used on the state field (in this example on conditions). Then the output of the functions get combined (appended) and a new state with the combined field is returned. 
This is useful, especially for experimentalists. For instance, imagine first getting a random pool of conditions and then wanting to get a few conditions from different experimentalists. With this, we can get a random pool of n conditions, a conditions from experimentalist A, b conditions from experimentalist B, c conditions from experimentalist C, ... Without this function, the experimentalists would be working in serial like a pipeline (pool -> experimentalist A -> experimentalist B -> experimentalist C), with this function every experimentalist gets is input from the pool and the output is the combined conditions of all of them (a + b + c)

Reply:
gotcha, that makes sense. thanks for the context!